### DISCUSSION TITLE
Controller components

### DISCUSSION THREAD

**Discussion by data-handler** [redom/redom#447] [Ideas]
Hi,

I have used redom for some time and really like it.

One question I have is around using a separate class for controller and view, so as to keep the view as 'dumb' as possible.

However, if I want to add an event listener to the view when it mounts and remove the event listener when the view unmounts, I currently make the controller's 'el' the view and mount the controller. Is there a better way?

```
export default class MyView {
   constructor() {
        this.el = el('',
             this.someElement = el('.some-element')
        );
    }
};

export default class MyController {
    constructor() {
        this.el = new MyView();
    }

    onmount = () => {
        this.el.someElement.addEventListener('click', this.onClick);
    };

    onunmount = () => {
        this.el.someElement.removeEventListener('click', this.onClick);
    };

    onClick = (e) => {
        // do stuff
    };
};

mount(document.body, new MyController());
```

**Comment by pakastin** [comment]
You could maybe extend View class from Controller class:

class View extends Controller {
  ...

**Comment by data-handler** [comment]
Not sure how that would give the Controller access to the view? Maybe the other way around:

class Controller extends View

**Reply by pakastin** [reply]
I tend to create separate API class, that i pass around to Views.. I also create single simple update function for it, which schedules single animation frame to render the app:
```js
class Api {
  constructor () {
    ...
  }
  update () {
    if (!this.rendering) {
      this.rendering = requestAnimationFrame(() => {
        this.rendering = null;
        
        this.app.update(this);
      });
    }
  }
}

class App {
  constructor ({ api }) {
    this.el = el('.app',
      this.router = router(routes, { api }),
      ...
    );
  }
  update (api) {
    const { section } = api;
    this.router.update(section, api);
  }
}

const api = new Api();
const app = new App({ api });

api.app = app;
```

**Reply by pakastin** [reply]
But extending should make it possible for both Classes to access each other if you want to use MVC pattern..

**Reply by data-handler** [reply]
It turns out there's a much simpler way, at least, for my purposes

```
class Controller {
    constructor() {
        this.view = new View();
        this.view.onmount = this.onmount;
    }

    onmount = () => {
        console.log('Controller->onmount');
    };
}
```

**Reply by pakastin** [reply]
Looks good! There's no right or wrong here, RE:DOM is very flexible since there's no black magic.. 👍

**Reply by data-handler** [reply]
Hi, yes. I wasn't trying to say this is the 'best' solution, just that it works well for me.

Thank you for your help 👍
